source/common/http/http2/codec_impl.cc

```diff
@@ -147,6 +135,42 @@ void ConnectionImpl::StreamImpl::encodeHeadersBase(const HeaderMap& headers, boo
   parent_.sendPendingFrames();
 }
 
+void ConnectionImpl::ClientStreamImpl::encodeHeaders(const RequestHeaderMap& headers,
+                                                     bool end_stream) {
+  // This must exist outside of the scope of isUpgrade as the underlying memory is
+  // needed until submitHeaders has been called.
+  std::vector<nghttp2_nv> final_headers;
+  Http::RequestHeaderMapPtr modified_headers;
+  if (Http::Utility::isUpgrade(headers)) {
```

Comment:
I think I'd vote for pulling semantic changes into a separate PR which @alyssawilk can review monday.

I think I feel comfortable fully reviewing everything everything else.

Reply:
https://github.com/envoyproxy/envoy/pull/10283